sandbox/apparmor/notify/listener/listener.go

```diff
@@ -0,0 +1,368 @@
+
+	// ErrListenerNotSupported indicates that the kernel does not support apparmor prompting
+	ErrListenerNotSupported = errors.New("kernel does not support apparmor notifications")
+
+	osOpen      = os.Open
+	notifyIoctl = notify.Ioctl
+)
+
+// Request is a high-level representation of an apparmor prompting message.
+//
+// Each request must be replied to by writing a boolean to the YesNo channel.
+type Request struct {
+	// Pid is the identifier of the process triggering the request.
+	Pid uint32
+	// Label is the apparmor label on the process triggering the request.
+	Label string
+	// SubjectUID is the UID of the subject that triggered the prompt
+	SubjectUid uint32
+
+	// Path is the path of the file, as seen by the process triggering the request.
+	Path string
+	// Permission is the opaque permission that is being requested.
+	Permission interface{}
+	// YesNo is a channel for writing the response.
+	YesNo chan bool
+}
+
+func newRequest(msg *notify.MsgNotificationFile) (*Request, error) {
+	var perm interface{}
+	if msg.Class == notify.AA_CLASS_FILE {
+		_, missingPerms, err := msg.DecodeFilePermissions()
+		if err != nil {
+			return nil, err
+		}
+		perm = missingPerms
+	}
+	return &Request{
+		Pid:        msg.Pid,
+		Label:      msg.Label,
+		Path:       msg.Name,
+		SubjectUid: msg.SUID,
+
+		Permission: perm,
+
+		YesNo: make(chan bool, 1),
+	}, nil
+}
+
+// Listener contains low-level components for receiving notification requests
+// and responding with notification responses.
+type Listener struct {
+	// reqs is a channel with incoming requests. Each request is asynchronous
+	// and needs to be replied to.
+	reqs chan *Request
+	// errs is a channel for receiving asynchronous error messages from
+	// concurrently running parts of the listener system.
+	errs chan error
+
+	notifyFile *os.File
+	poll       *epoll.Epoll
+
+	tomb tomb.Tomb
+
+	// status keeps track of whether the listener has been run and/or closed,
+	// both to ensure that Run() and Close() are executed at most once each,
+	// and to ensure that the listener cannot be run after it has been closed.
+	// Must be read/modified atomically, and can only be changed in one of the
+	// following ways:
+	// - statusReady -> statusRunning
+	// - statusReady -> statusClosed
+	// - statusRunning -> statusClosed
+	status uint32
+}
+
+const (
+	statusReady uint32 = iota
+	statusRunning
+	statusClosed
+)
+
+// Register opens and configures the apparmor notification interface.
+//
+// If the kernel does not support the notification mechanism the error is ErrNotSupported.
+func Register() (*Listener, error) {
+	path := notify.SysPath
+	if override := os.Getenv("PROMPT_NOTIFY_PATH"); override != "" {
+		path = override
+	}
+
+	notifyFile, err := osOpen(path)
+	if err != nil {
+		if os.IsNotExist(err) {
+			return nil, ErrListenerNotSupported
+		}
+		return nil, fmt.Errorf("cannot open %q: %v", path, err)
+	}
+	defer func() {
+		if err != nil {
```

Comment:
I think according to your explanation, all assignments of `err` in `Register()` after that initial `defer` should be safe to work with that `defer`, as they're all in the same scope (afaict). Please correct me if I'm wrong though.

It's true, in line 150 `err` is assigned within an if-statement, but it's using `if err = poll.Register(...); err != nil {`, which still assigns the existing `err` variable, rather than declaring a new one.

It is all a bit risky though, in multiple places if a particular non-nil error is returned by a called function and assigned to `err`, then the `Register()` function itself returns a different non-nil error. In every case though, a non-nill `err` corresponds to a non-nil return value from `Register()`, so it should be fine.

I'm very open to more elegant solutions though, if you have any suggestions.